### ISSUE TITLE
Migrations not being run after upgrade to Rails 8.0

### ISSUE THREAD

**Issue by thimo** [rails/rails#53752]
After upgrading our Rails 7.2 application to Rails 8.0 I can't run migrations anymore, and I'm a bit lost as to why this is happening. I've since gone back in commits to right before the 8.0 upgrade and migrations run just fine. In a newly generated Rails 8.0 project migrations run as well. I'm wondering is there's something missing in my setup that's causing this, any pointers to what it could be?

### Steps to reproduce
On our upgraded Rails 8.0 application:
```
$ rails g migration TestMigration
      invoke  active_record
      create    db/migrate/20241126161240_test_migration.rb
$ rails db:migrate
$
```

Going back to the last commit before gems where updated:
```
$ rails g migration TestMigration
      invoke  active_record
      create    db/migrate/20241126161525_test_migration.rb
$ rails db:migrate
== 20241126161525 TestMigration: migrating ====================================
== 20241126161525 TestMigration: migrated (0.0000s) ===========================

$
```

Not sure if I can provide more details around our app. In a newly generated Rails 8.0 app everything is fine. 

### Expected behavior
I would expect the migration to be run.

### Actual behavior
Migration does not run, no changes in the migration are applied.

### System configuration
**Rails version**: 8.0.0

**Ruby version**: 3.3.6

**Comment by WillNigel23** [comment]
This is happening to me too and is blocking my CI specs from running.

Checking the changelogs, I found this: https://github.com/rails/rails/blob/8-0-stable/activerecord/CHANGELOG.md

Particularly this:

```
When running db:migrate on a fresh database, load the databases schemas before running migrations.
```

Seems like it pre-loads `db/schema` first. Clearing/deleting the contents of this file and running db:migrate now makes it run the migrations again. 

Though this seems more like a bug instead of a feature. Are there flags to disable this? (especially since there usually are CRUD operations being done in some migration files)

**Comment by skipkayhil** [comment]
> Seems like it pre-loads `db/schema` first. Clearing/deleting the contents of this file and running db:migrate now makes it run the migrations again.

I believe you're seeing something different, this issue appears to be about migrations that were _just generated_ not being run.

> especially since there usually are CRUD operations being done in some migration files

I highly recommend not doing this, as its very brittle and easy to break.

---

> Not sure if I can provide more details around our app. In a newly generated Rails 8.0 app everything is fine.

Without a reproduction I'm afraid it will be difficult to provide any insight. Since you are able to reproduce locally I recommend `bundle open`ing gems and using `debug` to step through execution.

**Comment by WillNigel23** [comment]
It may still be relevant, as in my experience running `db:migrate` produces no logs like this when the schema version is the latest migration file (even after db:drop db:create)
```
== 20241126161525 TestMigration: migrating ====================================
== 20241126161525 TestMigration: migrated (0.0000s) ===========================
```

I created a branch that outlines the reproduction steps. Hopefully this helps: https://github.com/WillNigel23/migration_bug_test


Also some more trace logs when `db/schema` already exists
```
› rails db:migrate --trace
** Invoke db:migrate (first_time)
** Invoke db:load_config (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute db:load_config
** Execute db:migrate
** Invoke db:_dump (first_time)
** Execute db:_dump
** Invoke db:schema:dump (first_time)
** Invoke db:load_config
** Execute db:schema:dump
```

@thimo could you also please share the logs from `db:migrate --trace`? Thanks!

**Comment by WillNigel23** [comment]
The issue I am describing seems to be related to this #53649

**Comment by thimo** [comment]
@WillNigel23 here's the output for `rails db:migrate --trace`:
```
$ rails g migration TestMigration
      invoke  active_record
      create    db/migrate/20241127114910_test_migration.rb
$ rails db:migrate --trace
** Invoke db:migrate (first_time)
** Invoke db:load_config (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute db:load_config
** Execute db:migrate
** Invoke db:_dump (first_time)
** Execute db:_dump
** Invoke db:schema:dump (first_time)
** Invoke db:load_config
** Execute db:schema:dump
** Invoke environment
```

> Without a reproduction I'm afraid it will be difficult to provide any insight. Since you are able to reproduce locally I recommend bundle opening gems and using debug to step through execution.

@skipkayhil sure thing, do you maybe have a starting point for debugging? I'd assume somewhere in `ActiveRecord`, but what would be a good place to start?

**Comment by WillNigel23** [comment]
@thimo Please check https://github.com/rails/rails/issues/53649 too. 

One last thing to try running `rails db:migrate:reset` instead of just `rails db:migrate`, if it works then I think it is safe to say this is now a duplicate issue of the above and can be closed.

**Comment by skipkayhil** [comment]
> @skipkayhil sure thing, do you maybe have a starting point for debugging? I'd assume somewhere in `ActiveRecord`, but what would be a good place to start?

This is the entrypoint for the task: https://github.com/rails/rails/blob/2ab0094ab6e94d30a5a424dc93c2aeb1066cc76b/activerecord/lib/active_record/railties/databases.rake#L89

**Comment by thimo** [comment]
Thanks @skipkayhil for the starting point for debugging. 🙏 I found out that I had an initializer that overwrites the `ActiveRecord::Tasks::DatabaseTasks.migrate` method to increase database statement timeout for migrations. In Rails 8.0 an additional parameter was added to the method call (`skip_initialize: false`). I need a cleaner way to implement this overwrite or come up with a different approach to increase the DB statement timeout... 🧐

```
module AddStatementTimeoutToMigration
  def self.included base
    orig_method = base.method(:migrate)
    base.define_singleton_method :migrate do |*args|
      sql_statement = "SET statement_timeout = '#{10 * 60}s'"
      Rails.logger.debug { "-- #{sql_statement}" }
      ActiveRecord::Base.connection.execute(sql_statement)

      orig_method.call(*args)
    end
  end
end
ActiveRecord::Tasks::DatabaseTasks.send(:include, AddStatementTimeoutToMigration)
```

**Comment by yoav-epstein** [comment]
we needed to upgrade the `strong_migrations` gem.

**Comment by onerinas** [comment]
Thanks @yoav-epstein Upgrading the `strong_migrations` gem worked for me as well. If anyone interested, it was fixed in `strong_migrations` version 2.0.2: https://github.com/ankane/strong_migrations/blob/master/CHANGELOG.md#202-2024-10-30
